src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverter.cs

```diff
@@ -0,0 +1,255 @@
+                Index index, 
+                bool indexImplicitNull = false, 
+                bool indexEmptyEntries = true, 
+                string keyFieldName = null, 
+                bool storeValue = false, 
+                string storeValueFieldName = Constants.Documents.Indexing.Fields.ReduceKeyValueFieldName)
+        {
+            _index = index ?? throw new ArgumentNullException(nameof(index));
+            _index = index;
+            var fields = index.Definition.IndexFields.Values;
+            var dictionary = new Dictionary<string, IndexField>(fields.Count, default(OrdinalStringStructComparer));
+            foreach (var field in fields)
+                dictionary[field.Name] = field;
+            _fields = dictionary;
+            _blittableTraverser = storeValue ? BlittableJsonTraverser.FlatMapReduceResults : BlittableJsonTraverser.Default;
+            _allocator = new ByteStringContext(SharedMultipleUseFlag.None);
+            _knownFields = GetKnownFields();
+        }
+
+        public Dictionary<Slice, int> GetKnownFields()
+        {
+            if (_knownFields != null)
+                return _knownFields;
+            var knownFields = new Dictionary<Slice, int>();
+            foreach (var field in _fields.Values)
+            {
+                if(field.Name == null) 
+                    continue;
+                
+                Slice.From(_allocator, field.Name, ByteStringType.Immutable, out var value);
+                knownFields.Add(value, field.Id);
+            }
+
+            return knownFields;
+        }
+        
+        public Span<byte> InsertDocumentFields(LazyStringValue key, LazyStringValue sourceDocumentId, object doc, JsonOperationContext indexContext, out string id)
+        {
+            var document = (Document)doc;
+            _allocator.Allocate(document.Data.Size + _fields.Count * 1024, out ByteString buffer);
+            var entryWriter = new IndexEntryWriter(buffer.ToSpan(), _knownFields);
+            id = document.LowerId.ToLower();
+            bool shouldSkip;
+            foreach (var indexField in _fields.Values)
+            {
+                var scope = new CoraxWriterScope(_fields.Count);
+                if (BlittableJsonTraverserHelper.TryRead(_blittableTraverser, document, indexField.OriginalName ?? indexField.Name, out var value) == true)
+                {
+                    InsertRegularField(indexField, value, indexContext, out shouldSkip, ref entryWriter, scope);
+                }
+            }
+
+            //todo maciej try to figure out how should I deal with shouldSkip
+            entryWriter.Finish(out var output);
+            _allocator.Release(ref buffer);
+            return output;
+        }
+        
+        private void InsertRegularField(IndexField field, object value, JsonOperationContext indexContext, out bool shouldSkip, ref IndexEntryWriter entryWriter, CoraxWriterScope scope, bool nestedArray = false)
```

Comment:
Just in case anybody else is going to review it. The code here is gonna be refactored